disco_aws_automation/disco_elasticsearch_archive.py

```diff
@@ -0,0 +1,498 @@
+        """
+        threshold = float(self.get_es_option("archive_threshold"))
+        if threshold > 1.0:
+            raise RuntimeError("ElasticSearch archive threshold cannot exceed 1")
+        logging.info("Using threshold: %s", threshold)
+        indices_to_delete = self._indices_to_delete(threshold)
+
+        if indices_to_delete:
+            for index in indices_to_delete:
+                logging.info("Deleting index (%s) from cluster (%s).",
+                             index, self.cluster_name)
+                if not dry_run:
+                    self.es_client.indices.delete(index)
+        else:
+            logging.info("No need to delete any indices.")
+
+    def snapshots(self, snapshots=None):
+        """
+        List snapshots their state & etc.
+        """
+        snaps = self.es_client.snapshot.get(self.repository_name, snapshots or '_all')
+        return snaps['snapshots']
+
+    def snapshot_state(self, snapshot):
+        """
+        Return state of specified snapshot or 'unknown'.
+        """
+        try:
+            for snap in self.snapshots(snapshot):
+                if snap['snapshot'] == snapshot:
+                    return snap['state']
+        except NotFoundError:
+            pass
+        return 'unknown'
+
+    def _wait_for_snapshot(self, snapshot):
+        """
+        Wait for specified snapshots to complete snapshotting process.
+        """
+        max_time = time() + SNAPSHOT_WAIT_TIMEOUT
+        while True:
+            sleep(SNAPSHOT_POLL_INTERVAL)
+            snap_state = self.snapshot_state(snapshot)
+            if snap_state in ES_SNAPSHOT_FINAL_STATES:
+                return snap_state
+            if time() > max_time:
+                raise TimeoutError(
+                    "Timed out ({0}s) waiting for {1} to enter final state."
+                    .format(SNAPSHOT_WAIT_TIMEOUT, snapshot)
+                )
+
+    def restore(self, begin_date, end_date, dry_run=False):
+        """
+        Bring back indices within the specified date range (inclusive) from archive to ES cluster
+        """
+        date_pattern = re.compile(r'^[\d]{4}(\.[\d]{2}){2}$')
+        if not re.match(date_pattern, begin_date):
+            raise RuntimeError("Invalid begin date (yyyy.mm.dd): {0}".format(begin_date))
+        if not re.match(date_pattern, end_date):
+            raise RuntimeError("Invalid end date (yyyy.mm.dd): {0}".format(end_date))
+
+        snapshots = [snap['snapshot']
+                     for snap in self.snapshots()
+                     if snap['snapshot'][-10:] >= begin_date and
+                        snap['snapshot'][-10:] <= end_date]
+        logging.debug("Snapshots within the specified date range: %s", snapshots)
+        if not snapshots:
+            logging.info("No snapshots within date range are found.")
+            return
+
+        existing_indices = self._archivable_indices()
```

Comment:
`existing_indices` is being used here so that we don't try to restore indices that are already in the cluster. I don't think we are looking the expiration dates here.